### DISCUSSION TITLE
class에서 클로저와 비슷한 상태관리를 어떻게 만들 수 있을까 ?

### DISCUSSION THREAD

**Discussion by D5ng** [React-Core-Learn/Web-Component#3] [논의]
정답이 아니더라도 본인이 생각하는 것들을 공유해주세요!!!

**Comment by D5ng** [comment]
다시 생각해보니, 클래스의 setter를 사용하면 state의 값을 변경할 수 있도록하면 되지않을까? 라는 생각이 떠올랐어요. 그리고 `this.state`를 읽기 전용이나 protected로 설정해주면 이 데이터를 직접적으로 수정을 할 수 없게되니 더 좋지않을까? 라는 생각이 났어요

정확히는 클로저와 완전히 다른 개념이지만 클로저의 특징인 특정 메서드로만 상태를 변화시키는 관점은 동일하다는 생각이 들었어요

즉 getter와 setter을 사용하면 데이터를 안전하게 보호함과 동시에 수정할 수 있지 않을까? 라는 생각입니다.

**Reply by Seoin02** [reply]
입력하고 있었는데 내용이 눈앞에서 바뀌었어요!

**Reply by D5ng** [reply]
수정하고 있었어요 ㅋㅋㅋㅋㅋㅋ

**Comment by Seoin02** [comment]
클로저는 중첩함수에서 외부 함수의 생명주기가 끝나도 외부 함수의 변수를 계속 참조할 수 있는 것을 의미합니다. 음... useState가 클로저를 기반으로 구현돼있다고 알고 있는데, useState를 클래스형으로 구현해보면 되지 않을까요? 근데 우리가 만든게 useState를 구현한게 아닌가요? 어렵습니다

**Reply by D5ng** [reply]
저희가 만든건 useState의 의사코드라고 생각해요. 저도 깊게 들어가니 머리가 아픈데, 

결국 클로저를 사용하는 이유는 특정 변수를 캡슐화하는것과 특정 메서드를 제공해 제어하는 목적이라고 생각해요.

그래서 위와 같이 typescript와 class의 oop를 활용한다면 캡슐화와 특정 메서드로만 상태를 제어하는것이 가능하지 않을까? 라는 생각이 들었네요
